### PR TITLE
Updating permission when refreshing druid datasource

### PR DESCRIPTION
**Pull request by ShengyaoQian**

### CODE REVIEW

**Comment by mistercrunch** [comment]
@ShengyaoQian please fix the codeclimate issues

File: superset/models/helpers.py [hunk 1]
```diff
@@ -116,12 +116,48 @@ def __init__(  # noqa
         self.status = status
         self.error_message = error_message
 
+def merge_perm(sm, permission_name, view_menu_name, connection):
```
Comment (mistercrunch) [inline superset/models/helpers.py]: Same as
https://github.com/airbnb/superset/blob/f374345860f4a3e307ea1772b099ce038868aa71/superset/security.py#L70
?

Comment (ShengyaoQian) [inline superset/models/helpers.py]: Yes functionality-wise. But it doesn't seem to work with listener. I got an error saying "use connection level operation". So the merge function here uses connection.

**Comment by mistercrunch** [comment]
Why is it necessary to add the `database_name`  column to `tables`?

**Comment by ShengyaoQian** [comment]
@mistercrunch because I'm getting perm in post_delete. And when this function is called, the object is already in detached state, so I can't get the database relationship when it's detached. So I added database_name to get the name of the related database. If there's a better way to do this can you let me know?

**Comment by mistercrunch** [comment]
Seems like a bad idea to have database_name in two places. There are many scenarios where they might get out-of-sync. I'm surprised the object is detached at that point, but if it's the case you should emit a query and get an attached object and modify it.

**Comment by ShengyaoQian** [comment]
Yes, that's a good point. But I'm not sure what you mean by getting an attached object and modify. The object is already deleted when it reaches post_deletion. Another way I think I can do is overwriting the delete function instead of the post_delete so the object is not detached. And having all related permissions removed in delete function. Sounds good to you? @mistercrunch

**Comment by mistercrunch** [comment]
@ShengyaoQian there's also a pre-delete hook if that works better for what you are trying to do

**Comment by ShengyaoQian** [comment]
@mistercrunch I think I want to remove the perms after the datasource is removed. Pre delete will delete the perms first right? If deletion failed on deleting the datasource, the perms would have been deleted.

**Comment by ShengyaoQian** [comment]
@mistercrunch does this looks good to you?

**Comment by mistercrunch** [comment]
Instead of duplicating the `_delete` method everywhere, can't you just call `super` and do things before or after it?

**Comment by ShengyaoQian** [comment]
@mistercrunch refactored~

**Comment by ShengyaoQian** [comment]
@mistercrunch

**Comment by ShengyaoQian** [comment]
@mistercrunch

**Pull request merged**